src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -664,18 +673,10 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         Ok(ret)
     }
 
-    /// For a package, return all targets which are registered as dependencies
-    /// for that package.
-    pub fn dep_targets(&self, unit: &Unit<'a>) -> CargoResult<Vec<Unit<'a>>> {
-        if unit.profile.run_custom_build {
-            return self.dep_run_custom_build(unit)
-        } else if unit.profile.doc && !unit.profile.test {
-            return self.doc_deps(unit);
-        }
-
+    fn used_deps(&self, unit: &Unit<'a>) -> CargoResult<Vec<Unit<'a>>> {
```

Comment:
This seems like a somewhat unfortunate addition in the slew of already-too-many ways to calculate dependencies :(

I didn't quite follow [your explanation](https://github.com/rust-lang/cargo/pull/4469#issuecomment-327062863) earlier, mind detailing a bit more what was going on?

Reply:
Yeah I agree. At least this one is private only. The issue here is the fork on
`if unit.profile.run_custom_build` and `if unit.profile.doc && !unit.profile.test`

Specifically, OUT_DIR was set incorrectly in the build phase of doctests because doctests had a different dependency tree, but OUT_DIR was expected to be the same.

Reply:
Hm I'm still not 100% following... In any case, I'll check this out locally and poke around.

Reply:
Yeah. Easier if you poke around. I'll give it another shot though.

Without this refactor:
- When you compile the build.rs script for a doctest vs compiling the build.rs script for a regular test, it ends up with different metadata. This causes OUT_DIR to get set to a different (nonexistent) directory during doctests. It shows up as a test failure.
- https://travis-ci.org/rust-lang/cargo/jobs/271875681

With this refactor
- Doctests and regular tests have the same `used_deps` despite having different `dep_targets`.

Overall, it definitely does feel like some elements are repeated here and architecturally there is some unnecessary complexity, but I don't understand it well enough right now to find a way out. I **think** we need one function that returns deps for the package vs deps for the unit of build?

Reply:
Ah ok that sounds very surprising! The build script should be constant here and shouldn't change metadata, but let me poke around to see if I can't find where the difference is arising from